<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Комментарий ревьювера</h1>

Екатерина, привет! 

🤝 Тобой проделана большая работа, молодец! 

✏️ В этой тетрадке далее мои комментарии будут выделены по цветам:

    
<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Зеленный блок указывает на твои удачные решения и сильные стороны работы
</div>



    
<div class="alert alert-block alert-warning">
<b> ⚠️ Комментарий ревьюера:</b> Желтый блок указывает на то, что задание решено корректно, но какие-то вещи можно улучшить для большей эффективности работы, чистоты кода, поддержки и пр.
</div>


<div class="alert alert-block alert-danger">
<b> ❌ Комментарий ревьюера:</b> Красный блок указывает на то, то в задании допущена ошибка, которая влияет на итоговые результаты, и без исправления которой работа не будет принята.
</div>


    
📈 Не удаляй мои комментарии, так будет проще следить за динамикой.
    
🖋 Для простоты нашего общения и ускорения процесса проверки, можешь тоже выделять свои комментарии
<div class="alert alert-info"><b>Комментарий студента:</b> Например, вот так.</div>
    
  Ты проделала интересную работу. Ты справилась с большинством шагов, но по некоторым пунктам нужны доработки.
   Давай посмотрим, что получилось!

# SQL

**Цель**: предоставить данные, которые помогут сформулировать ценностное предложение для создания нового приложения для тех, кто любит читать.

**Кейс**: Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Подключение создано верно, база работает
</div>

Сохраним таблицы в переменные, чтобы использовать функцию для вывода результата через pandas.

In [2]:
books = '''SELECT * FROM books LIMIT 5'''
authors = '''SELECT * FROM authors LIMIT 5'''
publishers = '''SELECT * FROM publishers LIMIT 5'''
ratings = '''SELECT * FROM ratings LIMIT 5'''
reviews = '''SELECT * FROM reviews LIMIT 5'''

Функция для вывода:

In [3]:
def table(sql):
    return pd.io.sql.read_sql(sql, con = engine)

Посмотрим на первые 5 строк таблиц:

In [4]:
table(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
table(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
table(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
table(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-success">
    <b>✔️ Комментарий ревьюера:</b> Здорово, что, просматривая исследуемые таблицы, используешь метод LIMIT для ограничения вывода
</div>

## Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года;

In [8]:
books_cnt = '''
SELECT COUNT(book_id) 
FROM books
WHERE publication_date > '2000-01-01'
'''
table(books_cnt)

count
0    819

**819 книг** было выпущено после 1 января 2000 года

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Запрос написан верно, книги подсчитаны
</div>

## Задание 2: Для каждой книги посчитайте количество обзоров и среднюю оценку;

Для подсчёта рейтинга и количества обзоров необходимо присоединить таблицы с рейтингом ratings и обзорами reviews к таблице с книгами books

In [9]:
ratings_reviews = '''
SELECT books.title AS title,
       books.book_id AS book_id,
       AVG(ratings.rating) AS avg_rating,
       COUNT(DISTINCT reviews.text) AS reviews_cnt
FROM books
FULL OUTER JOIN reviews ON reviews.book_id = books.book_id
FULL OUTER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.title,
         books.book_id
ORDER BY reviews_cnt DESC, avg_rating DESC     

'''
table(ratings_reviews)

title  book_id  avg_rating  \
0                              Twilight (Twilight  #1)      948    3.662500   
1    Harry Potter and the Prisoner of Azkaban (Harr...      302    4.414634   
2    Harry Potter and the Chamber of Secrets (Harry...      299    4.287500   
3                                       The Book Thief      656    4.264151   
4                                     The Glass Castle      734    4.206897   
..                                                 ...      ...         ...   
995  Disney's Beauty and the Beast (A Little Golden...      191    4.000000   
996                               Leonardo's Notebooks      387    4.000000   
997                          Essential Tales and Poems      221    4.000000   
998    Anne Rice's The Vampire Lestat: A Graphic Novel       83    3.666667   
999                            The Natural Way to Draw      808    3.000000   

     reviews_cnt  
0              7  
1              6  
2              6  
3              6  
4              6  
..           ...  
995            0  
996            0  
997            0  
998            0  
999            0  

[1000 rows x 4 columns]

На первом месте по количеству обзоров **Twilight (Twilight #1)**, хотя рейтинг не самый высокий.

<div class="alert alert-block alert-danger">
<b> ❌ Комментарий ревьюера:</b> Ты посчитала всё верно, но книги без ревью и рейтинга тоже должны попадать в датасет, а сейчас у тебя они отсекаются.
</div>

<div class="alert alert-info"><b>Комментарий студента:</b> Скорректировала.</div>

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера 2:</b> Хорошо
</div>

## **Задание 3: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [10]:
publisher50 = '''
SELECT publishers.publisher AS name_publisher,
       COUNT(books.book_id) AS cnt__books
FROM publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY name_publisher
ORDER BY cnt__books DESC
LIMIT 1
'''
table(publisher50)

name_publisher  cnt__books
0  Penguin Books          42

Наибольшее количество книг более 50 страниц выпустило издательство **Penguin Books** - 42 книги.

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Отлично, у тебя получилось отфильровать данные по условиям
</div>

## Задание 4: Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:
author_high_rating = '''

WITH a AS
  (SELECT authors.author AS author,
          books.book_id AS id,
          AVG(ratings.rating) AS avg_rating,
          COUNT(ratings.rating) AS cnt_rating
FROM authors
INNER JOIN books ON books.author_id = authors.author_id
INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY author,
         id
HAVING COUNT(ratings.rating) >= 50)

SELECT a.author AS author,
       AVG(a.avg_rating) AS avg_rating
FROM a
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1    
'''
table(author_high_rating)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Автор с самой высокой средней оценкой книг (50 оценок и более) - **J.K. Rowling/Mary GrandPré**.

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Верное решение! 

## Задание 5 Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [12]:
avg_reviews = '''

WITH a AS
  (SELECT username,
          COUNT(rating_id) as cnt
   FROM ratings
   GROUP BY username), 
        
b AS 
  (SELECT COUNT(text)
   FROM reviews
   FULL OUTER JOIN a ON a.username = reviews.username
   WHERE cnt > 48
   GROUP BY reviews.username)        


SELECT ROUND(avg(count)) 
FROM b                                 
'''
table(avg_reviews)

round
0   24.0

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 шт.

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> Здорово, что ты умеешь писать временные таблицы и используешь округление :)
</div>

Общий вывод:
- после 1 января 2000 года вышло **819 книг**;
- в таблице **ratings_reviews** посчитан для каждой книги количество обзоров и средняя оценка: например, лидер по кол-ву обзоров Twilight  (Twilight #1). Согласно составленому рейтингу рекомендуется добавление популярных книг в приложение;
- издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**;
- автор с самой высокой средней оценкой книг(50 и более оценок) - **J.K. Rowling/Mary GrandPré**. Включение авторов с высокой средней оценкой книг также привлечет больше пользователей;
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок - **24**.
       

<div class="alert alert-block alert-success">
<b>✔️ Комментарий ревьюера:</b> В выводах отражены все необходимые результаты
</div>

<div class="alert alert-info"><b>Комментарий студента:</b> Спасибо за ревью!</div>